src/trainerai/main.c

```diff
@@ -5,25 +5,5515 @@
+given an attacker and a defender. The scores start at 100, and are modified accordingly based on the 
+the power, usefulness, or lackthereof of the move. If you want to add a flag, you can mirror the way
+these are written, and add it to the list of moveEvaluators. The flags are all defined in trainer_ai.h,
+so change those names there, too.
+*/
+
+/*BasicFlag:
+Heavily penalize stupid decisions that would fail, do nothing, or objectively hurt the user.*/
+
+int BasicFlag (struct BattleSystem *bsys, u32 attacker, int i, AiContext *ai){
+    int moveScore = 0;
+    struct BattleStruct *ctx = bsys->sp;
+
+    /*Never use moves that are out of pp*/
+    if(ctx->battlemon[ai->attacker].pp[i] == 0){
+        moveScore -= 40;
+    }
+
+    /*Check for ai->defender type immunities.*/
+    if(ai->attackerMoveEffectiveness == MOVE_STATUS_FLAG_NOT_EFFECTIVE){
+        moveScore -= 10;
+    }
+    /*Check for wonder guard*/
+    if(ai->attackerMoveEffectiveness != MOVE_STATUS_FLAG_SUPER_EFFECTIVE &&
+        ai->defenderAbility == ABILITY_WONDER_GUARD && ai->attackerAbility != ABILITY_MOLD_BREAKER){
+        moveScore -= 10;
+    }
+
+    //Ability immunities
+    if(ai->attackerAbility != ABILITY_MOLD_BREAKER){
+        if(ai->attackerMoveType == TYPE_ELECTRIC && //electric
+            (ai->defenderAbility == ABILITY_VOLT_ABSORB ||
+            ai->defenderAbility == ABILITY_LIGHTNING_ROD ||
+            ai->defenderAbility == ABILITY_MOTOR_DRIVE) ){
+            moveScore -= 10;
+        }
+        if(ai->attackerMoveType == TYPE_WATER && //water
+            (ai->defenderAbility == ABILITY_STORM_DRAIN ||
+            ai->defenderAbility == ABILITY_WATER_ABSORB ||
+            ai->defenderAbility == ABILITY_DRY_SKIN || 
+            ctx->field_condition & WEATHER_EXTREMELY_HARSH_SUNLIGHT)  ){
+            moveScore -= 10;
+        }
+        if(ai->attackerMoveType == TYPE_FIRE && //fire
+            (ai->defenderAbility == ABILITY_FLASH_FIRE || 
+            ai->defenderAbility == ABILITY_WELL_BAKED_BODY || 
+            ai->defenderAbility == ABILITY_THERMAL_EXCHANGE ||
+            ctx->field_condition & WEATHER_HEAVY_RAIN) ){
+            moveScore -= 10;
+        }
+        if(ai->attackerMoveType == TYPE_GRASS && //grass
+            (ai->defenderAbility == ABILITY_SAP_SIPPER) ){
+            moveScore -= 10;
+        }
+        if(ai->attackerMoveType == TYPE_GROUND && //ground
+            (ai->defenderAbility == ABILITY_LEVITATE ||
+            ai->defenderAbility == ABILITY_EARTH_EATER) ){
+            moveScore -= 10;
+        }
+        if(IsMoveSoundBased(ai->attackerMove) && //sound based moves
+            ai->defenderAbility == ABILITY_SOUNDPROOF){ 
+            moveScore -= 10;
+        }
+        if(ctx->moveTbl[ai->attackerMove].split == SPLIT_STATUS &&
+            ai->defenderAbility == ABILITY_GOOD_AS_GOLD){
+            moveScore -= 10;
+        }
```

Comment:
Edge cases not considered.
https://bulbapedia.bulbagarden.net/wiki/Good_as_Gold_(Ability)